cpp/src/arrow/array/builder_binary.h

```diff
@@ -77,56 +89,184 @@ class ARROW_EXPORT BinaryBuilder : public ArrayBuilder {
+          if (values[i]) {
+            value_data_builder_.UnsafeAppend(reinterpret_cast<const uint8_t*>(values[i]),
+                                             value_lengths[i]);
+          } else {
+            UnsafeAppendToBitmap(valid_bytes + valid_bytes_offset,
+                                 i - valid_bytes_offset);
+            UnsafeAppendToBitmap(false);
+            valid_bytes_offset = i + 1;
+          }
+        }
+      }
+      UnsafeAppendToBitmap(valid_bytes + valid_bytes_offset, length - valid_bytes_offset);
+    } else {
+      if (have_null_value) {
+        std::vector<uint8_t> valid_vector(length, 0);
+        for (int64_t i = 0; i < length; ++i) {
+          UnsafeAppendNextOffset();
+          if (values[i]) {
+            value_data_builder_.UnsafeAppend(reinterpret_cast<const uint8_t*>(values[i]),
+                                             value_lengths[i]);
+            valid_vector[i] = 1;
+          }
+        }
+        UnsafeAppendToBitmap(valid_vector.data(), length);
+      } else {
+        for (int64_t i = 0; i < length; ++i) {
+          UnsafeAppendNextOffset();
+          value_data_builder_.UnsafeAppend(reinterpret_cast<const uint8_t*>(values[i]),
+                                           value_lengths[i]);
+        }
+        UnsafeAppendToBitmap(NULLPTR, length);
+      }
+    }
+    return Status::OK();
+  }
+
+  void Reset() override {
+    ArrayBuilder::Reset();
+    offsets_builder_.Reset();
+    value_data_builder_.Reset();
+  }
+
+  Status Resize(int64_t capacity) override {
+    if (capacity > kListMaximumElements) {
+      return Status::CapacityError(
+          "BinaryBuilder cannot reserve space for more than 2^31 - 1 child elements, "
+          "got ",
+          capacity);
+    }
+    ARROW_RETURN_NOT_OK(CheckCapacity(capacity, capacity_));
+
+    // One more than requested for offsets
+    ARROW_RETURN_NOT_OK(offsets_builder_.Resize(capacity + 1));
+    return ArrayBuilder::Resize(capacity);
+  }
 
   /// \brief Ensures there is enough allocated capacity to append the indicated
   /// number of bytes to the value data buffer without additional allocations
-  Status ReserveData(int64_t elements);
+  Status ReserveData(int64_t elements) {
+    const int64_t size = value_data_length() + elements;
+    ARROW_RETURN_IF(size > memory_limit(),
+                    Status::CapacityError("Cannot reserve capacity larger than ",
+                                          memory_limit(), " bytes"));
+
+    return (size > value_data_capacity()) ? value_data_builder_.Reserve(elements)
+                                          : Status::OK();
+  }
 
-  Status FinishInternal(std::shared_ptr<ArrayData>* out) override;
+  Status FinishInternal(std::shared_ptr<ArrayData>* out) override {
+    // Write final offset (values length)
+    ARROW_RETURN_NOT_OK(AppendNextOffset());
 
-  /// \cond FALSE
-  using ArrayBuilder::Finish;
-  /// \endcond
+    // These buffers' padding zeroed by BufferBuilder
```

Comment:
not for this change, but I am kind of curious why we made zero out in buffer builder always versus only zero-ing for the bitmap (which I think is the only thing that requires it).

Reply:
I don't really know... Ideally zeroing the padding should be cheap operation, but perhaps not if you resize often.

Reply:
I would have to look I think when we resize we pad everything from size to capacity.  If capacity doubles size each time, it seems like it is effectively an O(N) cost?


anyways for a separate PR.

Reply:
That's possible. Can you open a JIRA?